conda/gateways/connection/adapters/jlap/adapter.py

```diff
@@ -0,0 +1,118 @@
+# -*- coding: utf-8 -*-
+# Copyright (C) 2012 Anaconda, Inc
+# SPDX-License-Identifier: BSD-3-Clause
+
+"""
+Cached repodata with `.jlap` incremental diffs, implemented as a ConnectionAdapter.
+"""
+
+import gzip
+import json
+import logging
+
+from requests import PreparedRequest, Response
+from requests.adapters import HTTPAdapter
+
+from conda.base import constants as consts
+from .repodata_proxy import RepodataUrl, fetch_repodata_json, static_file_headers, patch_files
+from .sync_jlap import SyncJlap
+
+log = logging.getLogger(__name__)
+
+
+class JlapAdapter(HTTPAdapter):
+    def __init__(self, base_adapter, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.base_adapter = base_adapter
+
+    def send(self, request, **kwargs):
+        """
+        Overrides the default HTTPAdapter. When this particular HTTPAdapter is in
+        use, we only care about intercepting URLs that end with
+        ``conda.base.constants.REPODATA_FN``
+        """
+        if request.url.endswith(consts.REPODATA_FN):
+            log.debug("Intercept %s", request.url)
+
+            # Pre-processing to the request, including changing its URL
+            repodata_url = RepodataUrl(request.url)
+            sync = SyncJlap(repodata_url)
+            new_request = self.prepare_request(request, sync)
+            response = self.send(new_request)
+
+            # This means there was an error with the response. We return it immediately.
+            if response.status_code > 300:
+                return response
+
+            # This step does some saving to the filesystem plus other processing
+            response = self.post_response_processing(response, sync)
+
+            # Post-processing to the response, including changing back its URL
+            return self.prepare_response(response, request)
+        else:
+            log.debug("Skip intercept %s", request.url)
+            return self.base_adapter.send(request, **kwargs)
```

Comment:
Hmm, so I guess the `JlapAdapter` is implemented so that it can wrap any of the other adapters? Hence why we use `base_adapter` instead of `super`.